src/Vision/VisionHelpersTrait.php

```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Vision;
+
+use Google\ApiCore\ArrayTrait;
+use InvalidArgumentException;
+
+/**
+ * Provides helper methods for generated Vision clients.
+ */
+trait VisionHelpersTrait
+{
+    use ArrayTrait;
+
+    /**
+     * A list of allowed url schemes.
+     *
+     * @var array
+     */
+    private $urlSchemes = [
+        'http',
+        'https',
+        'gs'
+    ];
+
+    private function annotateImageHelper(
+        $callback,
+        $requestClass,
+        $image,
+        $features,
+        $optionalArgs = []
+    ) {
+
+        $request = new $requestClass();
+        $request->setImage($image);
+        $request->setFeatures($features);
+        $imageContext = $this->pluck('imageContext', $optionalArgs, false);
+        if (!is_null($imageContext)) {
+            $request->setImageContext($imageContext);
+        }
+        return $callback([$request], $optionalArgs)->getResponses()[0];
+    }
+
+    private function buildFeatureList(
+        $featureClass,
+        $featureTypes
+    ) {
+
+        $features = [];
+        foreach ($featureTypes as $featureType) {
+            $feature = new $featureClass();
+            $feature->setType($featureType);
+            $features[] = $feature;
+        }
+        return $features;
+    }
+
+    private function createImageHelper($imageClass, $imageSourceClass, $imageInput)
+    {
+        $image = new $imageClass();
+        if (is_string($imageInput)) {
+            if (in_array(parse_url($imageInput, PHP_URL_SCHEME), $this->urlSchemes)) {
+                $imageSource = new $imageSourceClass();
+                $imageSource->setImageUri($imageInput);
+                $image->setSource($imageSource);
+            } else {
+                $image->setContent($imageInput);
```

Comment:
I assume PHP still has no reasonable distinction between text strings and byte strings?

It would be nice if you could detect filenames here. Ideally, if someone sent `'/path/to/image.jpeg'`, you could read the file from disk and call `$image->setContent` on _that_.

This is worthwhile to have even if we have to approximate it....a couple potential options:

  * do a regex test against `^[\w\d/\\._ %-]$+`
  * assume that any string with a length below 1024 (?) is a filename and not a binary image.

Reply:
This is a really interesting idea. My concern are those edge cases where something might fall through the cracks and not be handled correctly. Balancing that vs. just always asking users call `fopen` or `file_get_contents`, I lean towards the latter as it is what we already do throughout the library elsewhere. This seems like a good thing to take a vote on, though!

Reply:
I went with the current options because they are used elsewhere in the library. I would not be opposed to checking for this, but feel like the benefit may not be worth the complexity given that reading the file contents is a one-liner.